packages/jaeger-ui/src/model/ddg/DdgEVManager.tsx

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { compareKeys } from './visibility-key';
+
+import { PathElem, DdgEdge, DdgVertex, TDdgModel } from './types';
+
+export default class DdgEVManager {
+  private edges: Set<DdgEdge>;
+  private lastVisibilityKey: string;
+  private pathElemToVertex: Map<PathElem, DdgVertex>;
+  private vertices: Map<string, DdgVertex>;
+  private visibilityIdxToPathElem: Map<number, PathElem>;
+
+  constructor({ ddgModel }: { ddgModel: TDdgModel }) {
+    this.edges = new Set();
+    this.lastVisibilityKey = '';
+    this.pathElemToVertex = new Map();
+    this.vertices = new Map();
+    this.visibilityIdxToPathElem = ddgModel.visibilityIdxToPathElem;
+  }
+
+  private addElems = (newIndices: number[]) => {
+    newIndices.forEach(newIdx => {
+      const pathElem = this.visibilityIdxToPathElem.get(newIdx);
+      if (!pathElem) {
+        throw new Error(`Given visibilityIdx: "${newIdx}" that does not exist`);
+      }
+      const key = this.getVertexKey(pathElem);
+
+      let vertex = this.vertices.get(key);
+      if (!vertex) {
+        vertex = new DdgVertex({ key });
+        this.vertices.set(key, vertex);
+      }
+
+      this.pathElemToVertex.set(pathElem, vertex);
+      vertex.pathElems.add(pathElem);
+
+      const connectedElem = pathElem.focalSideNeighbor;
+      if (connectedElem) {
+        const connectedVertex = this.pathElemToVertex.get(connectedElem);
+        if (!connectedVertex) {
+          throw new Error(`Non-focal pathElem lacks connectedVertex. PathElem: ${pathElem}`);
+        }
+
+        if (!vertex[pathElem.focalSideEdgesKey].has(connectedVertex)) {
+          const newEdge = new DdgEdge(
+            pathElem.focalSideEdgesKey === 'ingressEdges'
+              ? {
+                  from: connectedVertex,
+                  to: vertex,
+                }
+              : {
+                  from: vertex,
+                  to: connectedVertex,
+                }
+          );
+          vertex[pathElem.focalSideEdgesKey].set(connectedVertex, newEdge);
```

Comment:
If the connected vertex is on the edge, why use it as a key to the edge?

Reply:
Using the vertices as keys makes it very easy to check if vertex `ƒ` has another vertex `A` as an ingress or egress edge (`ƒ.egressEdges.has(A)` or `ƒ.ingressEdges.has(A)`, respectively). It also makes it easy to remove edges when hiding `PathElem`s.
The other simple alternatives would be to either have:
1) a `Set<TEdge>` that would require checking if a specific vertexKey exists as a `to`/`from` on a set entry
or
2) a `Set<string>` of vertex keys to check if the vertexKey exists and then look it up in `DdgEVManager.edges`